### ISSUE TITLE
Docker amd64 unable to build with ruby 3.3.0 - nokogiri 1.16.0 fails to install with libjemalloc2 libjemalloc-dev debian deps installed

### ISSUE THREAD

**Issue by mandros1985** [jemalloc/jemalloc#2595]
Please tell me what can I fix in my dockerfile to make this build with jemalloc 

my local dev.Dockerfile 

```
FROM ruby:3.3.0-bookworm

ENV APP_HOME=/app \
#    LD_PRELOAD="libjemalloc.so.2" \
#    MALLOC_CONF="dirty_decay_ms:1000,narenas:2,background_thread:true,stats_print:true" \
    RUBY_YJIT_ENABLE="1" \
    LANG=C.UTF-8 \
    RUBY_VERSION=3.3.0 \
    BUNDLER_VERSION=2.5.4 \
    CHAMBER_VERSION=2.13.6 \
    NODE_VERSION=20 \
    BUNDLE_PATH=/bundle \
    BUNDLE_BIN=/bundle/bin \
    GEM_HOME=/bundle \
    RAILS_ENV=development \
    BUNDLE_GITLAB__COM=change_me \
    BUNDLE_GITHUB__COM=change_me
#libjemalloc2 libjemalloc-dev
# dose not work with jemalloc deps - had to remove it for now TODO: fix it

RUN apt-get update -qq \
    && DEBIAN_FRONTEND=noninteractive apt-get install -y wget nano zlib1g-dev liblzma-dev patch pkg-config

RUN wget --quiet -O - https://www.postgresql.org/media/keys/ACCC4CF8.asc | apt-key add -
RUN echo "deb http://apt.postgresql.org/pub/repos/apt bookworm-pgdg main" |tee  /etc/apt/sources.list.d/pgdg.list

RUN apt-get update -qq \
    && curl -fsSL https://deb.nodesource.com/setup_$NODE_VERSION.x | bash - \
    && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends nodejs

RUN apt-get update -qq \
    && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends libxml2-dev nano libxslt-dev git tzdata unzip openssl postgresql-client libjpeg62 libpng-dev libxrender1 libxext-dev wkhtmltopdf libfontconfig1 libssl-dev

RUN wget -q -O /bin/chamber https://github.com/segmentio/chamber/releases/download/v$CHAMBER_VERSION/chamber-v${CHAMBER_VERSION}-linux-amd64
RUN chmod +x /bin/chamber \
    && rm -rf /var/lib/apt/lists/* \
    && sed -i 's,^\(MinProtocol[ ]*=\).*,\1'TLSv1.0',g' /etc/ssl/openssl.cnf \
    && sed -i 's,^\(CipherString[ ]*=\).*,\1'DEFAULT@SECLEVEL=1',g' /etc/ssl/openssl.cnf

# install latest chrome for testing purposes
RUN wget -q -O - https://dl-ssl.google.com/linux/linux_signing_key.pub | apt-key add - \
    && echo 'deb http://dl.google.com/linux/chrome/deb/ stable main' >> /etc/apt/sources.list.d/google.list
RUN apt-get update && apt-get -y install google-chrome-stable

#RUN wget -q http://archive.ubuntu.com/ubuntu/pool/main/o/openssl/openssl_1.1.1f-1ubuntu2.20_amd64.deb \
#    && yes | dpkg -i ./openssl_1.1.1f-1ubuntu2.20_amd64.deb
# old deps - left fpr refference

RUN mkdir ${APP_HOME}
WORKDIR ${APP_HOME}
RUN gem install bundler:$BUNDLER_VERSION
RUN bundle config set --global force_ruby_platform true
COPY Gemfile $APP_HOME/
COPY Gemfile.lock $APP_HOME/
RUN bundle install -j40 --retry 30 \
    && bundle check || bundle install
COPY . ${APP_HOME}
```

**Comment by mandros1985** [comment]
using also docker compose 

```
version: "3.9"
services:
  app:
    platform: linux/x86_64
    build:
      context: .
      
      ... TLDR

```

**Comment by mandros1985** [comment]
When build as is the nokogiri inside the image is 

```

root@c830f01b9add:/app# bundle exec nokogiri --version
# Nokogiri (1.16.0)
    ---
    warnings: []
    nokogiri:
      version: 1.16.0
      cppflags:
      - "-I/bundle/ruby/3.3.0/gems/nokogiri-1.16.0/ext/nokogiri"
      - "-I/bundle/ruby/3.3.0/gems/nokogiri-1.16.0/ext/nokogiri/include"
      - "-I/bundle/ruby/3.3.0/gems/nokogiri-1.16.0/ext/nokogiri/include/libxml2"
      ldflags: []
    ruby:
      version: 3.3.0
      platform: x86_64-linux
      gem_platform: x86_64-linux
      description: ruby 3.3.0 (2023-12-25 revision 5124f9ac75) +YJIT [x86_64-linux]
      engine: ruby
    libxml:
      source: packaged
      precompiled: false
      patches:
      - 0001-Remove-script-macro-support.patch
      - 0002-Update-entities-to-remove-handling-of-ssi.patch
      - 0003-libxml2.la-is-in-top_builddir.patch
      - '0009-allow-wildcard-namespaces.patch'
      - 0010-update-config.guess-and-config.sub-for-libxml2.patch
      - 0011-rip-out-libxml2-s-libc_single_threaded-support.patch
      memory_management: ruby
      iconv_enabled: true
      compiled: 2.12.3
      loaded: 2.12.3
    libxslt:
      source: packaged
      precompiled: false
      patches:
      - 0001-update-config.guess-and-config.sub-for-libxslt.patch
      datetime_enabled: true
      compiled: 1.1.39
      loaded: 1.1.39
    other_libraries:
      libgumbo: 1.0.0-nokogiri
```

**Comment by mandros1985** [comment]
when building with the jemalloc deps it throws 

```
662.0 --- End jemalloc statistics ---
662.0 
662.0 make failed, exit code 2
662.0 
662.0 Gem files will remain installed in /bundle/ruby/3.3.0/gems/nokogiri-1.16.0 for
662.0 inspection.
662.0 Results logged to
662.0 /bundle/ruby/3.3.0/extensions/x86_64-linux/3.3.0/nokogiri-1.16.0/gem_make.out
662.0 
662.0   /usr/local/lib/ruby/3.3.0/rubygems/ext/builder.rb:125:in `run'
662.0   /usr/local/lib/ruby/3.3.0/rubygems/ext/builder.rb:51:in `block in make'
662.0   /usr/local/lib/ruby/3.3.0/rubygems/ext/builder.rb:43:in `each'
662.0   /usr/local/lib/ruby/3.3.0/rubygems/ext/builder.rb:43:in `make'
662.0   /usr/local/lib/ruby/3.3.0/rubygems/ext/ext_conf_builder.rb:42:in `build'
662.0   /usr/local/lib/ruby/3.3.0/rubygems/ext/builder.rb:193:in `build_extension'
662.0 /usr/local/lib/ruby/3.3.0/rubygems/ext/builder.rb:227:in `block in
662.0 build_extensions'
662.0   /usr/local/lib/ruby/3.3.0/rubygems/ext/builder.rb:224:in `each'
662.0   /usr/local/lib/ruby/3.3.0/rubygems/ext/builder.rb:224:in `build_extensions'
662.0   /usr/local/lib/ruby/3.3.0/rubygems/installer.rb:852:in `build_extensions'
662.0 /bundle/gems/bundler-2.5.4/lib/bundler/rubygems_gem_installer.rb:76:in
662.0 `build_extensions'
662.0 /bundle/gems/bundler-2.5.4/lib/bundler/rubygems_gem_installer.rb:28:in
662.0 `install'
662.0   /bundle/gems/bundler-2.5.4/lib/bundler/source/rubygems.rb:205:in `install'
662.0 /bundle/gems/bundler-2.5.4/lib/bundler/installer/gem_installer.rb:54:in
662.0 `install'
662.0 /bundle/gems/bundler-2.5.4/lib/bundler/installer/gem_installer.rb:16:in
662.0 `install_from_spec'
662.0 /bundle/gems/bundler-2.5.4/lib/bundler/installer/parallel_installer.rb:132:in
662.0 `do_install'
662.0 /bundle/gems/bundler-2.5.4/lib/bundler/installer/parallel_installer.rb:123:in
662.0 `block in worker_pool'
662.0   /bundle/gems/bundler-2.5.4/lib/bundler/worker.rb:62:in `apply_func'
662.0 /bundle/gems/bundler-2.5.4/lib/bundler/worker.rb:57:in `block in
662.0 process_queue'
662.0   <internal:kernel>:187:in `loop'
662.0   /bundle/gems/bundler-2.5.4/lib/bundler/worker.rb:54:in `process_queue'
662.0 /bundle/gems/bundler-2.5.4/lib/bundler/worker.rb:90:in `block (2 levels) in
662.0 create_threads'
662.0 
662.0 An error occurred while installing nokogiri (1.16.0), and Bundler cannot
662.0 continue.
662.0 
662.0 In Gemfile:
662.0   rails was resolved to 7.1.3, which depends on
662.0     actioncable was resolved to 7.1.3, which depends on
662.0       actionpack was resolved to 7.1.3, which depends on
662.0         actionview was resolved to 7.1.3, which depends on
662.0           rails-dom-testing was resolved to 2.2.0, which depends on
662.0             nokogiri
662.5 ___ Begin jemalloc statistics ___
662.5 Version: "5.3.0-0-g54eaed1d8b56b1aa528be3bdd1877e59c56fa90c"

664.7 Fetching gem metadata from https://rubygems.org/.........
666.2 Installing nokogiri 1.16.0 with native extensions
1322.8 Gem::Ext::BuildError: ERROR: Failed to build gem native extension.
1322.8 
1322.8     current directory: /bundle/ruby/3.3.0/gems/nokogiri-1.16.0/ext/nokogiri
1322.8 /usr/local/bin/ruby extconf.rb
1322.8 checking for whether -std=c99 is accepted as CFLAGS... no
1322.8 checking for whether -Wno-declaration-after-statement is accepted as CFLAGS...
1322.8 no
1322.8 checking for whether -O2 is accepted as CFLAGS... no
1322.8 checking for whether -g is accepted as CFLAGS... no
1322.8 checking for whether -Winline is accepted as CFLAGS... no
1322.8 checking for whether -Wmissing-noreturn is accepted as CFLAGS... no
1322.8 checking for whether -Wconversion -Wno-sign-conversion is accepted as CFLAGS...
1322.8 no
1322.8 Building nokogiri using packaged libraries.
1322.8 Static linking is enabled.
1322.8 Cross build is disabled.
1322.8 Using mini_portile version 2.8.5
1322.8 checking for iconv... yes
1322.8 ---------- IMPORTANT NOTICE ----------
1322.8 Building Nokogiri with a packaged version of libxml2-2.12.3.
1322.8 Configuration options: --host\=x86_64-pc-linux --enable-static --disable-shared
1322.8 --libdir\=/bundle/ruby/3.3.0/gems/nokogiri-1.16.0/ports/x86_64-linux/libxml2/2.12.3/lib
1322.8 --with-iconv\=yes --disable-dependency-tracking --without-python
1322.8 --without-readline --with-c14n --with-debug --with-threads --without-tls
1322.8 --disable-shared --enable-static CPPFLAGS\= CFLAGS\=-O2\ -U_FORTIFY_SOURCE\ -g\
1322.8 -fPIC
1322.8 The following patches are being applied:
1322.8   - 0001-Remove-script-macro-support.patch
1322.8   - 0002-Update-entities-to-remove-handling-of-ssi.patch
1322.8   - 0003-libxml2.la-is-in-top_builddir.patch
1322.8   - 0009-allow-wildcard-namespaces.patch
1322.8   - 0010-update-config.guess-and-config.sub-for-libxml2.patch
1322.8   - 0011-rip-out-libxml2-s-libc_single_threaded-support.patch
1322.8 
1322.8 The Nokogiri maintainers intend to provide timely security updates, but if
1322.8 this is a concern for you and want to use your OS/distro system library
1322.8 instead, then abort this installation process and install nokogiri as
1322.8 instructed at:
1322.8 
1322.8 https://nokogiri.org/tutorials/installing_nokogiri.html#installing-using-standard-system-libraries
1322.8 
1322.8 Note, however, that nokogiri cannot guarantee compatibility with every
1322.8 version of libxml2 that may be provided by OS/package vendors.
1322.8 
1322.8 Extracting libxml2-2.12.3.tar.xz into
1322.8 tmp/x86_64-pc-linux/ports/libxml2/2.12.3... OK
1322.8 Running git apply with
1322.8 /bundle/ruby/3.3.0/gems/nokogiri-1.16.0/patches/libxml2/0001-Remove-script-macro-support.patch...
1322.8 OK
1322.8 Running git apply with
1322.8 /bundle/ruby/3.3.0/gems/nokogiri-1.16.0/patches/libxml2/0002-Update-entities-to-remove-handling-of-ssi.patch...
1322.8 OK
1322.8 Running git apply with
1322.8 /bundle/ruby/3.3.0/gems/nokogiri-1.16.0/patches/libxml2/0003-libxml2.la-is-in-top_builddir.patch...
1322.8 OK
1322.8 Running git apply with
1322.8 /bundle/ruby/3.3.0/gems/nokogiri-1.16.0/patches/libxml2/0009-allow-wildcard-namespaces.patch...
1322.8 OK
1322.8 Running git apply with
1322.8 /bundle/ruby/3.3.0/gems/nokogiri-1.16.0/patches/libxml2/0010-update-config.guess-and-config.sub-for-libxml2.patch...
1322.8 OK
1322.8 Running git apply with
1322.8 /bundle/ruby/3.3.0/gems/nokogiri-1.16.0/patches/libxml2/0011-rip-out-libxml2-s-libc_single_threaded-support.patch...
1322.8 OK
1322.8 Running 'configure' for libxml2 2.12.3... OK
1322.8 Running 'compile' for libxml2 2.12.3... OK
1322.8 Running 'install' for libxml2 2.12.3... OK
1322.8 Activating libxml2 2.12.3 (from
1322.8 /bundle/ruby/3.3.0/gems/nokogiri-1.16.0/ports/x86_64-linux/libxml2/2.12.3)...
1322.8 Using mini_portile version 2.8.5
1322.8 ---------- IMPORTANT NOTICE ----------
1322.8 Building Nokogiri with a packaged version of libxslt-1.1.39.
1322.8 Configuration options: --host\=x86_64-pc-linux --enable-static --disable-shared
1322.8 --libdir\=/bundle/ruby/3.3.0/gems/nokogiri-1.16.0/ports/x86_64-linux/libxslt/1.1.39/lib
1322.8 --disable-dependency-tracking --without-python --without-crypto --with-debug
1322.8 --with-libxml-prefix\=/bundle/ruby/3.3.0/gems/nokogiri-1.16.0/ports/x86_64-linux/libxml2/2.12.3
1322.8 --disable-shared --enable-static CFLAGS\=-O2\ -U_FORTIFY_SOURCE\ -g\ -fPIC
1322.8 The following patches are being applied:
1322.8   - 0001-update-config.guess-and-config.sub-for-libxslt.patch
1322.8 
1322.8 The Nokogiri maintainers intend to provide timely security updates, but if
1322.8 this is a concern for you and want to use your OS/distro system library
1322.8 instead, then abort this installation process and install nokogiri as
1322.8 instructed at:
1322.8 
1322.8 https://nokogiri.org/tutorials/installing_nokogiri.html#installing-using-standard-system-libraries
1322.8 
1322.8 Extracting libxslt-1.1.39.tar.xz into
1322.8 tmp/x86_64-pc-linux/ports/libxslt/1.1.39... OK
1322.8 Running git apply with
1322.8 /bundle/ruby/3.3.0/gems/nokogiri-1.16.0/patches/libxslt/0001-update-config.guess-and-config.sub-for-libxslt.patch...
1322.8 OK
1322.8 Running 'configure' for libxslt 1.1.39... OK
1322.8 Running 'compile' for libxslt 1.1.39... OK
1322.8 Running 'install' for libxslt 1.1.39... OK
1322.8 Activating libxslt 1.1.39 (from
1322.8 /bundle/ruby/3.3.0/gems/nokogiri-1.16.0/ports/x86_64-linux/libxslt/1.1.39)...
1322.8 checking for whether -DNOKOGIRI_PACKAGED_LIBRARIES is accepted as CPPFLAGS... no
1322.8 xml2-config cflags:
1322.8 -I/bundle/ruby/3.3.0/gems/nokogiri-1.16.0/ports/x86_64-linux/libxml2/2.12.3/include/libxml2
1322.8 checking for whether
1322.8 -DNOKOGIRI_LIBXML2_PATCHES="\"0001-Remove-script-macro-support.patch
1322.8 0002-Update-entities-to-remove-handling-of-ssi.patch
1322.8 0003-libxml2.la-is-in-top_builddir.patch 0009-allow-wildcard-namespaces.patch
1322.8 0010-update-config.guess-and-config.sub-for-libxml2.patch
1322.8 0011-rip-out-libxml2-s-libc_single_threaded-support.patch\"" is accepted as
1322.8 CPPFLAGS... no
1322.8 checking for -llzma... yes
1322.8 xslt-config cflags:
1322.8 -I/bundle/ruby/3.3.0/gems/nokogiri-1.16.0/ports/x86_64-linux/libxml2/2.12.3/include/libxml2
1322.8 -I/bundle/ruby/3.3.0/gems/nokogiri-1.16.0/ports/x86_64-linux/libxslt/1.1.39/include
1322.8 checking for whether
1322.8 -DNOKOGIRI_LIBXSLT_PATCHES="\"0001-update-config.guess-and-config.sub-for-libxslt.patch\""
1322.8 is accepted as CPPFLAGS... no
1322.8 checking for xmlParseDoc() in libxml/parser.h... yes
1322.8 checking for xsltParseStylesheetDoc() in libxslt/xslt.h... yes
1322.8 checking for exsltFuncRegister() in libexslt/exslt.h... yes
1322.8 Using mini_portile version 2.8.5
1322.8 ---------- IMPORTANT NOTICE ----------
1322.8 Building Nokogiri with a packaged version of libgumbo-1.0.0-nokogiri.
1322.8 Configuration options: --disable-shared --enable-static CFLAGS\=-fPIC
1322.8 Copying gumbo-parser files into
1322.8 tmp/x86_64-pc-linux/ports/libgumbo/1.0.0-nokogiri/gumbo-parser...
1322.8 Running 'compile' for libgumbo 1.0.0-nokogiri... OK
1322.8 Activating libgumbo 1.0.0-nokogiri (from
1322.8 ports/x86_64-linux/libgumbo/1.0.0-nokogiri)...
1322.8 checking for whether
1322.8 -I/bundle/ruby/3.3.0/gems/nokogiri-1.16.0/ext/nokogiri/ports/x86_64-linux/libgumbo/1.0.0-nokogiri/include
1322.8 is accepted as CPPFLAGS... no
1322.8 checking for gumbo_parse_with_options() in nokogiri_gumbo.h... yes
1322.8 checking for xmlHasFeature()... yes
1322.8 checking for xmlFirstElementChild()... yes
1322.8 checking for xmlRelaxNGSetParserStructuredErrors()... yes
1322.8 checking for xmlRelaxNGSetValidStructuredErrors()... yes
1322.8 checking for xmlSchemaSetValidStructuredErrors()... yes
1322.8 checking for xmlSchemaSetParserStructuredErrors()... yes
1322.8 checking for rb_category_warning()... yes
1322.8 checking for whether
1322.8 -DNOKOGIRI_OTHER_LIBRARY_VERSIONS="\"libgumbo:1.0.0-nokogiri\"" is accepted as
1322.8 CPPFLAGS... no
1322.8 creating Makefile


1322.8 
1322.8 current directory: /bundle/ruby/3.3.0/gems/nokogiri-1.16.0/ext/nokogiri
1322.8 make DESTDIR\= sitearchdir\=./.gem.20240117-40744-1974oq
1322.8 sitelibdir\=./.gem.20240117-40744-1974oq
1322.8 compiling gumbo.c
1322.8 gumbo.c:32:10: fatal error: nokogiri_gumbo.h: No such file or directory
1322.8    32 | #include "nokogiri_gumbo.h"
1322.8       |          ^~~~~~~~~~~~~~~~~~
1322.8 compilation terminated.


1322.8 --- End jemalloc statistics ---
1322.8 make: *** [Makefile:472: gumbo.o] Error 1
1322.8 ___ Begin jemalloc statistics ___



```

**Comment by mandros1985** [comment]
As it turned out, a simple upgrade to rails 7.1.3 from 7.1.2 was enough to make it work flawlessly ... 

Also removed redundant nokogiri arch variants from gemfile.lock - left only the x86_64 (had multiple) 

Even with the native bundler platform set for ruby, it worked ...
